compiler/rustc_middle/src/ty/consts/kind.rs

```diff
@@ -68,6 +73,53 @@ pub enum ConstKind<'tcx> {
     /// A placeholder for a const which could not be computed; this is
     /// propagated to avoid useless error messages.
     Error(ty::DelaySpanBugEmitted),
+
+    /// Expr which contains an expression which has partially evaluated items.
+    Expr(Expr<'tcx>),
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash)]
+#[derive(HashStable, TyEncodable, TyDecodable, TypeVisitable, TypeFoldable)]
+pub enum Expr<'tcx> {
+    Binop(mir::BinOp, Const<'tcx>, Const<'tcx>),
+    UnOp(mir::UnOp, Const<'tcx>),
+    FunctionCall(Const<'tcx>, &'tcx List<Const<'tcx>>),
+    Cast(CastKind, Const<'tcx>, Ty<'tcx>),
+}
+
+#[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
+static_assert_size!(Expr<'_>, 24);
+
+impl<'tcx> Expr<'tcx> {
+    pub fn unify_failure_kind(self) -> FailureKind {
```

Comment:
shouldn't be needed, should it?

we can now use the following on an expanded const `ct`:
```rust
if ct.has_infer_types_or_consts() {
    ...
} else if ct.has_param_types_or_consts() {
    ...
} else {
    ...
}
```
don't think we want to keep a separate method for that